File: csharp/HearthstoneReplays/Parser/ReplayParser.cs
#region

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Runtime.InteropServices;
using System.Text.RegularExpressions;
using HearthDb.Enums;
using HearthstoneReplays.Parser.Handlers;
using HearthstoneReplays.Parser.ReplayData;

#endregion

namespace HearthstoneReplays.Parser
{
	public class ReplayParser
	{
		public const string Version = "1.0";
		public const int HearthstoneBuild = 10833;
		private static readonly ParserState State = new ParserState();

		public static HearthstoneReplay FromFile(string filePath, int hsBuild = HearthstoneBuild, params GameType[] gameTypes)
		{
			using(var reader = new StreamReader(filePath))
				return FromTextReader(reader);
		}

		public static HearthstoneReplay FromTextReader(TextReader reader, int hsBuild = HearthstoneBuild, params GameType[] gameTypes)
		{
			var lines = new List<string>();
			string line;
			while((line = reader.ReadLine()) != null)
				lines.Add(line);
			return FromString(lines);
		}

		public static HearthstoneReplay FromString(IEnumerable<string> lines, int hsBuild = HearthstoneBuild, params GameType[] gameTypes)
		{
			Read(lines.ToArray());
			State.Replay.Version = Version;
			State.Replay.Build = hsBuild.ToString();
			for(var i = 0; i < State.Replay.Games.Count; i++)
			{
				if(gameTypes.Length == 1)
					State.Replay.Games[i].Type = (int)gameTypes[0];
				else
					State.Replay.Games[i].Type = gameTypes.Length > i ? (int)gameTypes[i] : 0;
			}
			return State.Replay;
		}

		public static void Read(string[] lines)
		{
			State.Reset();
			State.Replay.Games = new List<Game>();
			Regex logTypeRegex = null;
			foreach(var line in lines)
			{
				Match match;
				if(logTypeRegex == null)
				{
					match = Regexes.PowerlogLineRegex.Match(line);
					if(match.Success)
						logTypeRegex = Regexes.PowerlogLineRegex;
					else
					{
						match = Regexes.OutputlogLineRegex.Match(line);
						if(match.Success)
							logTypeRegex = Regexes.OutputlogLineRegex;
					}
				}
				else
					match = logTypeRegex.Match(line);
				if(!match.Success)
					continue;
				AddData(match.Groups[1].Value, match.Groups[2].Value, match.Groups[3].Value);
			}
		}

		private static void AddData(string timestamp, string method, string data)
		{
			switch(method)
			{
				case "GameState.DebugPrintPower":
					DataHandler.Handle(timestamp, data, State);
					break;
				case "GameState.SendChoices":
					SendChoicesHandler.Handle(timestamp, data, State);
					break;
				case "GameState.DebugPrintChoices":
				case "GameState.DebugPrintEntityChoices":
                    ChoicesHandler.Handle(timestamp, data, State);
					break;
				case "GameState.DebugPrintEntitiesChosen":
					EntityChosenHandler.Handle(timestamp, data, State);
					break;
                case "GameState.DebugPrintOptions":
					OptionsHandler.Handle(timestamp, data, State);
					break;
				case "GameState.SendOption":
					SendOptionHandler.Handle(timestamp, data, State);
					break;
				case "GameState.OnEntityChoices":
					// Spectator mode noise
					break;
				case "ChoiceCardMgr.WaitThenShowChoices":
					// Not needed for replays
					break;
				case "GameState.DebugPrintChoice":
					Console.WriteLine("Warning: DebugPrintChoice was removed in 10357. Ignoring.");
                    break;
				default:
					if(!method.StartsWith("PowerTaskList.") && !method.StartsWith("PowerProcessor.") && !method.StartsWith("PowerSpellController"))
						Console.WriteLine("Warning: Unhandled method: " + method);
					break;
			}
		}
	}
}

File: java/src/main/java/info/hearthsim/hsreplay/parser/ReplayParser.java
package info.hearthsim.hsreplay.parser;

import java.util.ArrayList;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import info.hearthsim.hsreplay.enums.GameType;
import info.hearthsim.hsreplay.parser.handlers.DataHandler;
import info.hearthsim.hsreplay.parser.handlers.OptionsHandler;
import info.hearthsim.hsreplay.parser.replaydata.Game;
import info.hearthsim.hsreplay.parser.replaydata.HearthstoneReplay;
import lombok.extern.slf4j.Slf4j;

@Slf4j
public class ReplayParser {

	public static final String VERSION = "1.0";
	public static final int HEARTHSTONE_BUILD = 10833;

	private final ParserState state = new ParserState();

	public HearthstoneReplay fromString(Iterable<String> lines, GameType... gameTypes) throws Exception {
		return fromString(lines, HEARTHSTONE_BUILD, gameTypes);
	}

	public HearthstoneReplay fromString(Iterable<String> lines, int hsBuild, GameType... gameTypes) throws Exception {
		read(lines);
		state.getReplay().setVersion(VERSION);
		state.getReplay().setBuild(String.valueOf(hsBuild));
		for (int i = 0; i < state.getReplay().getGames().size(); i++) {
			if (gameTypes.length == 1) {
				state.getReplay().getGames().get(i).setType(gameTypes[0].getIntValue());
			}
			else {
				state.getReplay().getGames().get(i).setType(gameTypes.length > i ? gameTypes[i].getIntValue() : 0);
			}
		}
		return state.getReplay();
	}

	private void read(Iterable<String> lines) throws Exception {
		state.reset();
		state.getReplay().setGames(new ArrayList<Game>());
		Pattern logTypeRegex = null;

		for (String line : lines) {

			log.debug("Considering input log line: " + line);
			Matcher match = null;
			if (logTypeRegex == null) {
				match = Regexes.PowerlogLineRegex.matcher(line);
				if (match.matches()) {
					log.debug("\tMatching PowerlogLineRegex");
					logTypeRegex = Regexes.PowerlogLineRegex;
				}
				else {
					match = Regexes.OutputlogLineRegex.matcher(line);
					if (match.matches()) {
						log.debug("\tMatching OutputLineRegex");
						logTypeRegex = Regexes.OutputlogLineRegex;
					}
					else {
						log.debug("\tNot matching any regex " + line);
					}
				}
			}
			else {
				match = logTypeRegex.matcher(line);
			}

			if (!match.matches()) {
				continue;
			}

			addData(match.group(1), match.group(2), match.group(3));
		}
	}

	private void addData(String timestamp, String method, String data) throws Exception {

		switch (method) {
			case "GameState.DebugPrintPower":
				DataHandler.handle(timestamp, data, state);
				break;
			// case "PowerTaskList.DebugPrintPower":
			// DataHandler.handle(timestamp, data, state);
			// break;
			// case "GameState.SendChoices":
			// SendChoicesHandler.handle(timestamp, data, state);
			// break;
			// case "GameState.DebugPrintChoices":
			// case "GameState.DebugPrintEntityChoices":
			// ChoicesHandler.handle(timestamp, data, state);
			// break;
			// case "GameState.DebugPrintEntitiesChosen":
			// EntityChosenHandler.handle(timestamp, data, state);
			// break;
			case "GameState.DebugPrintOptions":
				OptionsHandler.handle(timestamp, data, state);
				break;
			// case "GameState.SendOption":
			// SendOptionHandler.handle(timestamp, data, state);
			// break;
			// case "GameState.OnEntityChoices":
			// // Spectator mode noise
			// break;
			// case "ChoiceCardMgr.WaitThenShowChoices":
			// // Not needed for replays
			// break;
			// case "GameState.DebugPrintChoice":
			// System.out.println("Warning: DebugPrintChoice was removed in
			// 10357. Ignoring.");
			// break;
			default:
				if (!method.startsWith("PowerTaskList.") && !method.startsWith("PowerProcessor.")
						&& !method.startsWith("PowerSpellController")) {
					log.warn("Warning: Unhandled method: " + method);
				}
				break;
		}
	}
}
